### PR TITLE
fix(tagging): hook up tagging functionality

### PR DESCRIPTION
**Pull request by anthony-liddle**
## Goal

Hooking up the tagging functionality

## Todos:
- [x] Add tagging related packages
- [x] Add tagging connector
- [x] Move existing tagging components to components folder
- [x] Ensure typeahead works
- [x] Hook up saving of tags

## Implementation Decisions

Still need to complete the visual changes to this but am waiting for storybooks

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -20,11 +20,14 @@
...
   },
   "dependencies": {
+    "downshift": "^6.1.7",
     "js-base64": "^3.7.2",
     "linaria": "^3.0.0-beta.13",
+    "match-sorter": "^6.3.1",
     "prop-types": "^15.7.2",
     "react": "^17.0.2",
     "react-dom": "^17.0.2",
+    "react-input-autosize": "^3.0.0",
```
Comment (anthony-liddle) [inline package.json:30]: Reminder to run `npm i` once this gets merged~

File: src/common/api/saving/tags.js [hunk 2]
```diff
@@ -16,7 +16,7 @@ export function syncItemTags(id, tags, actionInfo) {
   return request({
     path: 'send/',
     data: {
-      actions: [{ action: 'tags_add', item_id: id, tags, ...actionInfo }]
+      actions: [{ action: 'tags_replace', item_id: id, tags, ...actionInfo }]
```
Comment (anthony-liddle) [inline src/common/api/saving/tags.js:19]: There was a bug where if you saved by accident you couldn't remove it. This fixes it! 🎉

File: src/common/helpers.js [hunk 3]
```diff
@@ -109,6 +109,7 @@ export function closeLoginPage() {
 
 export function deriveItemData(item) {
   return {
+    itemId: item?.item_id,
```
Comment (anthony-liddle) [inline src/common/helpers.js:112]: We need this for saving tags

File: src/components/chips/chips.js [hunk 4]
```diff
@@ -44,39 +42,40 @@ const chipItem = css`
...
-    border-color: ${$overcast};
+    border-color: #999;
   }
 `
 
 const chipItemActive = css`
-  background-color: ${$smoke};
-  padding: '2px 5px 2px 6px';
+  background-color: #ddd;
+  padding: 2px 5px 2px 6px;
```
Comment (anthony-liddle) [inline src/components/chips/chips.js:56]: Updated the styles here to have the right colors, still needs some darkMode adjustments and some final tweaks. Will be much easier once storybooks is in place.

File: src/components/item-preview/item-preview.js [hunk 5]
```diff
@@ -33,13 +34,13 @@ const previewWrapper = css`
...
+  return (title && publisher) ? (
     <div className={previewWrapper}>
       {thumbnail ? <img src={thumbnail} /> : null}
       <div>
         <h3>{title}</h3>
         <p>{publisher}</p>
       </div>
     </div>
-  )
+  ) : null
```
Comment (anthony-liddle) [inline src/components/item-preview/item-preview.js:45]: This prevents a flash of incomplete content.

File: src/components/tagging/suggestions/suggestions.js [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+      </ul>
+    </div>
+  )
+}
+
+Suggestions.propTypes = {
+  suggestions: PropTypes.array,
+  addTag: PropTypes.func,
+  used: PropTypes.array
+}
```
Comment (anthony-liddle) [inline src/components/tagging/suggestions/suggestions.js:80]: This was moved back here from `connectors` and revamped slightly.

File: src/components/tagging/tagging.js [hunk 7]
```diff
@@ -0,0 +1,255 @@
...
+  }),
+  storedTags: PropTypes.array,
+  activateTag: PropTypes.func,
+  deactivateTag: PropTypes.func,
+  deactivateTags: PropTypes.func,
+  addTag: PropTypes.func,
+  removeTag: PropTypes.func,
+  removeTags: PropTypes.func,
+  closePanel: PropTypes.func
+}
```
Comment (anthony-liddle) [inline src/components/tagging/tagging.js:255]: Moved here from `connectors` and revamped slightly.

File: src/components/tagging/taginput/taginput.js [hunk 8]
```diff
@@ -147,13 +150,13 @@ export const Taginput = ({
...
-    </InputWrapper>
+    </div>
   )
 }
 
-Taginput.propTypes = {
+TagInput.propTypes = {
   getInputProps: PropTypes.func,
   setBlur: PropTypes.func,
   setFocus: PropTypes.func,
```
Comment (anthony-liddle) [inline src/components/tagging/taginput/taginput.js:162]: Moved and renamed.

File: src/connectors/doorhanger/doorhanger.js [hunk 9]
```diff
@@ -47,7 +44,7 @@ const doorhangerStyle = css`
  * Injected doorhanger to show on save
  * @param   {string} saveStatus  current status of the item operations
  */
-export const Doorhanger = ({ saveStatus, storedTags, suggestedTags, itemPreview }) => {
+export const Doorhanger = ({ saveStatus, itemPreview, closePanel }) => {
```
Comment (anthony-liddle) [inline src/connectors/doorhanger/doorhanger.js:47]: These tagging objects have been moved to the connector.

File: src/connectors/tagging/tagging.js [hunk 10]
```diff
@@ -1,264 +1,130 @@
...
+      tags={tagsObject}
+      storedTags={storedTags}
+      addTag={addTag}
+      activateTag={activateTag}
+      deactivateTag={deactivateTag}
+      deactivateTags={deactivateTags}
+      removeTag={removeTag}
+      removeTags={removeTags}
+      closePanel={closePanel}
+    />
```
Comment (anthony-liddle) [inline src/connectors/tagging/tagging.js:128]: This is now a straight connector which loads the `Tagging` component. This houses the state management functionality. This diff is useless.

File: src/pages/background/settings.js [hunk 11]
```diff
@@ -16,17 +16,8 @@ export async function initOptions() {
...
-  // const recsEnabled = (await getSetting('on_save_recommendations')) || 'unset'
-  // if (recsEnabled === 'unset') {
-  //   await setSettings({ on_save_recommendations: true })
-  // }
 }
 
-// export async function setOnSaveRecs(tab, { isEnabled }) {
-//   await setSettings({ on_save_recommendations: isEnabled })
-// }
-
```
Comment (anthony-liddle) [inline src/pages/background/settings.js:29]: Cleanup!

File: src/pages/injector/app.js [hunk 12]
```diff
@@ -41,14 +34,6 @@ export const App = () => {
...
       }
 
-      // case ARCHIVE_ITEM_SUCCESS: {
-      //   return setSaveStatus('archived')
-      // }
-
-      // case ARCHIVE_ITEM_FAILURE: {
-      //   return setSaveStatus('archive_failed')
-      // }
-
```
Comment (anthony-liddle) [inline src/pages/injector/app.js:51]: This functionality is going away... cleanup! 🧹

File: src/pages/injector/app.js [hunk 13]
```diff
@@ -57,16 +42,6 @@ export const App = () => {
...
-      case UPDATE_STORED_TAGS: {
-        const { tags } = payload
-        return setStoredTags(tags)
-      }
-
-      case SUGGESTED_TAGS_SUCCESS: {
-        const { suggestedTags } = payload
-        return setSuggestedTags(suggestedTags)
-      }
-
```
Comment (anthony-liddle) [inline src/pages/injector/app.js:69]: Moved this to the tagging connector.

File: src/pages/injector/app.js [hunk 14]
```diff
@@ -91,6 +66,10 @@ export const App = () => {
     setSaveStatus('idle')
   }
 
+  const closePanel = () => {
+    setSaveStatus('idle')
+  }
+
```
Comment (anthony-liddle) [inline src/pages/injector/app.js:72]: So we can pass this down and close the panel when you click enter in the tagging input.

File: src/components/chips/chips.js (hunk 4 above)
Comment (wtfluckey) [reply src/components/chips/chips.js:56]: Love this. I think this is a much better starting place.

File: src/components/tagging/tagging.js [hunk 15]
```diff
@@ -0,0 +1,250 @@
...
+  const makeTagActive = (tag) => {
+    return activateTag({ tag })
+  }
+
+  const makeTagInactive = (tag) => {
+    return deactivateTag({ tag })
+  }
+
+  const makeTagsInactive = (blur) => {
+    if (!markedTags.length) return blur ? inputRef.current.blur() : null
```
Comment (wtfluckey) [inline src/components/tagging/tagging.js:127]: What is `blur` in this context?

Comment (collectedmind) [reply src/components/tagging/tagging.js:127]: Blur is when the input box loses focus.

**Review by wtfluckey — APPROVED** [review]
Phew 😅 This is really great. Thank you for adding comments, it made reviewing this a lot easier. I pulled down the branch and tested and see no issues there too. I had just one tiny question to better understand `blur`.

File: src/components/tagging/tagging.js (hunk 15 above)
Comment (anthony-liddle) [reply src/components/tagging/tagging.js:127]: Specifically, we pass this function to the input. If someone presses the ESCAPE key we call this function and pass in a `bool` for `blur` to make the input lose focus and "deactive' the tags in the input (which means the X disappears).

**Pull request merged**